graph/testgraph/testgraph.go

```diff
@@ -0,0 +1,964 @@
+	Edges() graph.Edges
+}
+
+// weightedEdgeLister is a graph that can return all its weighted edges.
+type weightedEdgeLister interface {
+	// WeightedEdges returns all the weighted edges of a graph.
+	WeightedEdges() graph.WeightedEdges
+}
+
+// matrixer is a graph that can return an adjacency matrix.
+type matrixer interface {
+	// Matrix returns the graph's adjacency matrix.
+	Matrix() mat.Matrix
+}
+
+// ReturnAllNodes tests the constructed graph for the ability to return all
+// the nodes it claims it has used in its construction. This is a check of
+// the Nodes method of graph.Graph and the iterator that is returned.
+func ReturnAllNodes(t *testing.T, b Builder) {
+	for _, test := range testCases {
+		g, want, _, _, _, ok := b(test.nodes, test.edges, test.self, test.absent)
+		if !ok {
+			continue
+		}
+
+		it := g.Nodes()
+		if !isValidIterator(it) {
+			t.Errorf("invalid iterator for test %q: got:%#v", test.name, it)
+			continue
+		}
+		var got []graph.Node
+		for it != nil && it.Next() {
+			got = append(got, it.Node())
+		}
+
+		sort.Sort(ordered.ByID(got))
+		sort.Sort(ordered.ByID(want))
+
+		if !reflect.DeepEqual(got, want) {
+			t.Errorf("unexpected nodes result for test %q:\ngot: %v\nwant:%v", test.name, got, want)
+		}
+	}
+}
+
+// ReturnNodeSlice tests the constructed graph for the ability to return all
+// the nodes it claims it has used in its construction using the NodeSlicer
+// interface. This is a check of the Nodes method of graph.Graph and the
+// iterator that is returned.
+func ReturnNodeSlice(t *testing.T, b Builder) {
+	for _, test := range testCases {
+		g, want, _, _, _, ok := b(test.nodes, test.edges, test.self, test.absent)
+		if !ok {
+			continue
+		}
+
+		it := g.Nodes()
+		if !isValidIterator(it) {
+			t.Errorf("invalid iterator for test %q: got:%#v", test.name, it)
+			continue
+		}
+		if it == nil {
+			continue
+		}
+		s, ok := it.(graph.NodeSlicer)
+		if !ok {
+			t.Errorf("invalid type for test %q: %T cannot return node slicer", test.name, g)
+			continue
+		}
+		got := s.NodeSlice()
+
+		sort.Sort(ordered.ByID(got))
+		sort.Sort(ordered.ByID(want))
+
+		if !reflect.DeepEqual(got, want) {
+			t.Errorf("unexpected nodes result for test %q:\ngot: %v\nwant:%v", test.name, got, want)
+		}
+	}
+}
+
+// NodeExistence tests the constructed graph for the ability to correctly
+// return the existence of nodes within the graph. This is a check of the
+// Node method of graph.Graph.
+func NodeExistence(t *testing.T, b Builder) {
+	for _, test := range testCases {
+		g, want, _, _, _, ok := b(test.nodes, test.edges, test.self, test.absent)
+		if !ok {
+			continue
+		}
+
+		seen := make(set.Nodes)
+		for _, exist := range want {
+			seen.Add(exist)
+			if g.Node(exist.ID()) == nil {
+				t.Errorf("missing node for test %q: %v", test.name, exist)
+			}
+		}
+		for _, ghost := range test.nonexist {
+			if seen.Has(ghost) {
+				// Do not check nodes that the graph says should
```

Comment:
The thinking was that sometimes a test case could be used (rather than skipped), but the input node would be either shifted or dropped. This allows that. The alternative is to just allow tests to be skipped if they do not conform to the invariants of the tested type.

Reply:
I've added a `t.Logf` for when the `Builder` rejects a test. I could do something similar here, or just remove this condition.

Reply:
Yes, at least logging this would be better but it still confuses that there could be a discrepancy between what the test case creator thinks and what the graph thinks about which nodes should exist in the graph.

Reply:
The comment is bad. It's not the graph thinking it should exist, it's the graph builder. The `seen` map is filled from the `want` slice returned by the builder. The builder may decide to shift a node for example, while keeping the general intent of the test the same.

Reply:
Ok.